File: reverse-bits/Solution190.java
// https://leetcode.com/problems/reverse-bits/

class Solution190 {

    // T: O(1), S: O(1)
    public int reverseBits(int num) {
        var res = 0;
        var lsb = 0;
        for (var i = 31; i >= 0; i--) {
            lsb = num & 1;
            lsb <<= i;
            res |= lsb;
            num >>= 1;
        }
        return res;
    }

    public static void main(String[] args) {
        var sol = new Solution190();
        var num1 = Integer.parseInt("00000010100101000001111010011100", 2);
        System.out.println(sol.reverseBits(num1)); // 964176192
    }

}

File: reverse-bits/Solution190.py
# https://leetcode.com/problems/reverse-bits/

class Solution190:

    # T: O(1), S: O(1)
    def reverseBits(self, num: int) -> int:
        res = 0
        for i in range(32):
            lsb = num & 1
            res = (res << 1) | lsb
            num >>= 1
        return res


def main() -> None:
    sol = Solution190()
    num1 = int("00000010100101000001111010011100", 2)
    print(sol.reverseBits(num1))  # 964176192


if __name__ == "__main__":
    main()
